### ISSUE TITLE
yarn run gatsby build stalles in run queries in workers

### ISSUE THREAD

**Issue by hellow554** [gatsbyjs/gatsby#35192]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

When unning `yarn run gatsby build` the progress stalls, here's a log:

```
success compile gatsby files - 0.543s
success load gatsby config - 0.018s
success load plugins - 0.154s
success onPreInit - 0.001s
success delete worker cache from previous builds - 0.002s
success initialize cache - 0.091s
success copy gatsby files - 0.098s
success Compiling Gatsby Functions - 0.404s
success onPreBootstrap - 0.421s
success createSchemaCustomization - 0.002s
success Checking for changed pages - 0.001s
success source and transform nodes - 0.037s
info Writing GraphQL type definitions to /tmp/tmp.XznRs8rQkV/.cache/schema.gql
success building schema - 0.340s
success createPages - 0.001s
success createPagesStatefully - 0.079s
info Total nodes: 20, SitePage nodes: 1 (use --verbose for breakdown)
success Checking for changed pages - 0.001s
success Cleaning up stale page-data - 0.003s
success onPreExtractQueries - 0.001s
success extract queries from components - 0.130s
success write out redirect data - 0.005s
success onPostBootstrap - 0.002s
info bootstrap finished - 5.077s
success write out requires - 0.009s
success Building production JavaScript and CSS bundles - 0.818s
success Building HTML renderer - 0.372s
success Execute page configs - 0.036s
success Caching Webpack compilations - 0.003s
[                            ]   0.001 s 0/1 0% run queries in workers
```

---

If you remove the (empty) `gatsby-config.js` file it "compiles" just fine.
`npm run build` also just works fine.

### Reproduction Link

https://github.com/hellow554/gatsby-yarn

### Steps to Reproduce

1. `yarn install`
2. `yarn run gatsby build`

### Expected Result

The build is succeful

### Actual Result

The build stalls

### Environment

```shell
System:
    OS: Linux 5.10 Ubuntu 21.10 21.10 (Impish Indri)
    CPU: (8) x64 Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
    Shell: 3.1.2 - /usr/local/bin/fish
  Binaries:
    Node: 14.19.1 - /tmp/xfs-13e4cf48/node
    Yarn: 3.2.0 - /tmp/xfs-13e4cf48/yarn
    npm: 6.14.16 - /usr/bin/npm
  Languages:
    Python: 3.9.7 - /usr/bin/python
```

### Config Flags

```
module.exports = {};
```

**Comment by pieh** [comment]
Hey @hellow554, we are looking into yarn@2+ issues and work-in-progress https://github.com/gatsbyjs/gatsby/pull/35194 might also solve this issue.

**Comment by pieh** [comment]
Just out of curiosity - are you using PnP mode of yarn@3? or the `node_modules` / `node_linker` mode? (not sure if it's relevant, so far I was debugging issue with yarn@3 PnP, but the issue might very well be just yarn@3 regardless of mode)

**Comment by hellow554** [comment]
> Just out of curiosity - are you using PnP mode of yarn@3? or the `node_modules` / `node_linker` mode? 

I'm using the default and IIRC it's using the pnp mode.
Not an expert though.

**Comment by pieh** [comment]
Just released `@next` version (`gatsby@4.11.0-next.4`) and I get it to build now:

```
➜  gatsby-yarn git:(main) ✗ yarn why gatsby
└─ tmp.XznRs8rQkV@workspace:.
   └─ gatsby@npm:4.11.0-next.4 [4ed86] (via npm:4.11.0-next.4 [4ed86])
➜  gatsby-yarn git:(main) ✗ yarn gatsby clean && yarn run gatsby build
info Deleting .cache, public, /Users/misiek/test/gatsby-yarn/node_modules/.cache/babel-loader,
/Users/misiek/test/gatsby-yarn/node_modules/.cache/terser-webpack-plugin
info Successfully deleted directories

success compile gatsby files - 0.642s
success load gatsby config - 0.025s
success load plugins - 0.186s
success onPreInit - 0.001s
success initialize cache - 0.144s
success copy gatsby files - 0.155s
success Compiling Gatsby Functions - 0.403s
success onPreBootstrap - 0.425s
success createSchemaCustomization - 0.002s
success Checking for changed pages - 0.001s
success source and transform nodes - 0.050s
info Writing GraphQL type definitions to /Users/misiek/test/gatsby-yarn/.cache/schema.gql
success building schema - 0.497s
success createPages - 0.001s
success createPagesStatefully - 0.127s
info Total nodes: 20, SitePage nodes: 1 (use --verbose for breakdown)
success Checking for changed pages - 0.008s
success onPreExtractQueries - 0.001s
success extract queries from components - 0.117s
success write out redirect data - 0.057s
success onPostBootstrap - 0.001s
info bootstrap finished - 6.609s
success write out requires - 0.014s
success Building production JavaScript and CSS bundles - 7.670s
success Building HTML renderer - 4.274s
success Execute page configs - 0.021s
success Caching Webpack compilations - 0.001s
success run queries in workers - 0.039s - 1/1 25.52/s
success Merge worker state - 0.024s
success Rewriting compilation hashes - 0.021s
success Writing page-data.json files to public directory - 0.014s - 1/1 70.86/s
success Building static HTML for pages - 5.490s - 1/1 0.18/s
success onPostBuild - 0.001s
info Done building in 24.699195745 sec


Pages

└ src/pages/index.tsx
  └   /

  ╭────────────────────────────────────────────────────────────────╮
  │                                                                │
  │   (SSG) Generated at build time                                │
  │ D (DSG) Deferred static generation - page generated at runtime │
  │ ∞ (SSR) Server-side renders at runtime (uses getServerData)    │
  │ λ (Function) Gatsby function                                   │
  │                                                                │
  ╰────────────────────────────────────────────────────────────────╯
```

For now just in `@next`, but will be cherry-picking and releasing fix to `latest` line as well later today

**Comment by hellow554** [comment]
Can confirm it works. Thanks for the (very) fast fix! Amazing :heart:
